### PR TITLE
Add proper support for a single RequestTracingFilter that supports Servlet 2.x and Servlet 3 environments at the same time

### PR DESCRIPTION
**Pull request by nicmunroe**
Fixes issue #50. Big thanks to @adriancole and @woldie for the help on this.

### CODE REVIEW

File: testonly/testonly-old-servlet/src/test/java/com/nike/wingtips/componenttest/RequestTracingFilterOldServletComponentTest.java [hunk 1]
```diff
@@ -55,7 +58,30 @@
     private SpanRecorder spanRecorder;
 
     @BeforeClass
+    @SuppressWarnings("JavaReflectionMemberAccess")
     public static void beforeClass() throws Exception {
+        try {
+            ServletRequest.class.getMethod("getAsyncContext");
```
Comment (adriancole) [inline testonly/testonly-old-servlet/src/test/java/com/nike/wingtips/componenttest/RequestTracingFilterOldServletComponentTest.java]: Smart

File: wingtips-servlet-api/README.md [hunk 2]
```diff
@@ -72,13 +87,16 @@ See the [base project readme's section on propagation](../README.md#propagating_
...
-API is on the classpath at runtime, however if you receive `NoClassDefFoundError`s related to Servlet API classes then 
-you'll need to pull a Servlet API dependency into your project that supports a minimum Servlet 3.0.1 version. For 
-reference, `wingtips-servlet-api` uses the compile-only Servlet API dependency 
-[`javax.servlet:javax.servlet-api:3.0.1`](http://search.maven.org/#artifactdetails%7Cjavax.servlet%7Cjavax.servlet-api%7C3.0.1%7Cjar). 
+API is on the classpath at runtime, however if you receive class-not-found errors related to Servlet API classes then 
+you'll need to pull a Servlet API dependency into your project. Library authors who wish to build on functionality in
+this module might need to do this. Which Servlet API dependency you pull in depends on the type of Servlet environment 
+you want to support (Servlet 2.x or Servlet 3+). For example:
+
+* Servlet 3 API dependency: [`javax.servlet:javax.servlet-api:$servlet3ApiVersion`](http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22javax.servlet%22%20AND%20a%3A%22javax.servlet-api%22)
```
Comment (adriancole) [inline wingtips-servlet-api/README.md]: Variables intentional?

File: wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilter.java [hunk 3]
```diff
@@ -1,121 +1,199 @@
...
-     * introduced with the Servlet 3 API.
+     * This attribute key will be set to a value of true via {@link ServletRequest#setAttribute(String, Object)} the
+     * first time this filter's distributed tracing logic is run for any given request. This filter will then see this
+     * attribute on any subsequent executions for the same request and continue the filter chain without executing the
+     * distributed tracing logic again to make sure this filter's logic is only executed once per request.
      *
-     * @param servletRequestClass ServletRequest implementation class
-     * @return true if the given servletRequest implementation class supports the getAsyncContext() method,
-     * otherwise false
+     * <p>If you want to prevent this filter from executing on specific requests then you can override {@link
```
Comment (adriancole) [inline wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilter.java]: Neat.. good idea to document this

File: wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilter.java [hunk 4]
```diff
@@ -1,121 +1,199 @@
...
+            request.setAttribute(TraceHeaders.TRACE_ID, newSpan.getTraceId());
+            request.setAttribute(TraceHeaders.SPAN_ID, newSpan.getSpanId());
+            request.setAttribute(TraceHeaders.PARENT_SPAN_ID, newSpan.getParentSpanId());
+            request.setAttribute(TraceHeaders.SPAN_NAME, newSpan.getSpanName());
+            request.setAttribute(Span.class.getName(), newSpan);
+
+            // Make sure we set the trace ID on the response header now before the response is committed (if we wait
+            //      until after the filter chain then the response might already be committed, silently preventing us
+            //      from setting the response header)
+            response.setHeader(TraceHeaders.TRACE_ID, newSpan.getTraceId());
```
Comment (adriancole) [inline wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilter.java]: Ps eventhough b3 doesnt usually do this, I noticed amazon do. For example returning back the sample result (as everything is in one header)

File: wingtips-servlet-api/README.md (hunk 2 above)
Comment (nicmunroe) [reply wingtips-servlet-api/README.md]: Yes, but I can see how it looks like a copy/paste error. I'll see if I can clear it up.

File: wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilter.java (hunk 4 above)
Comment (nicmunroe) [reply wingtips-servlet-api/src/main/java/com/nike/wingtips/servlet/RequestTracingFilter.java]: Yeah I still have issue #30 to change this to have the response header be something other than `X-B3-TraceId`, but having the trace ID somewhere in the response headers has been critically useful.

**Pull request merged**